### PR TITLE
Mmap: fix grow! for non file IOs

### LINKED ISSUE #54203: Mmap.mmap regression from Julia 1.10.0 to 1.10.1
I use the Raspberry PI for some work.  This includes using the Package BaremetalPi to set pins high or low.  This all works well with Julia 1.10.0 but not with 1.10.1.  This is explained in more detail on discourse [here](https://discourse.julialang.org/t/dev-gpiomem-error-for-baremetalpi-mmap-mmap-regression/112611).  

The problem lies in Mmap as explained in Post 6 of the discourse link and repeated here.  All commands were given on a Raspberry PI 4 with 8 Gig RAM.  Julia was installed using Juliaup.

The output from 1.10.0 which is working is:

```
julia> versioninfo()
Julia Version 1.10.0
Commit 3120989f39b (2023-12-25 18:01 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (aarch64-linux-gnu)
  CPU: 4 Ã Cortex-A72
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, cortex-a72)
  Threads: 1 on 4 virtual cores

julia> using Mmap

julia> io = open("/dev/gpiomem", "w+")
IOStream(<file /dev/gpiomem>)

julia> gpiomap = Mmap.mmap(io, Vector{UInt32}, (1024,), 0, grow=false)
1024-element Vector{UInt32}:
 0x21200900
 0x00040264
 0x12040000
 0x3fffffff
 0x00000064
 0x00000000
 0x00000000
 0x6770696f
 0x6770696f
 0x6770696f
          â®
 0x6770696f
 0x6770696f
 0x6770696f
 0x6770696f
 0x6770696f
 0x6770696f
 0x6770696f
 0x6770696f
 0x6770696f
```

And from 1.10.1 which errors is:

```
julia> versioninfo()
Julia Version 1.10.1
Commit 7790d6f0641 (2024-02-13 20:41 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (aarch64-linux-gnu)
  CPU: 4 Ã Cortex-A72
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, cortex-a72)
Threads: 1 default, 0 interactive, 1 GC (on 4 virtual cores)

julia> using Mmap

julia> io = open("/dev/gpiomem", "w+")
IOStream(<file /dev/gpiomem>)

julia> gpiomap = Mmap.mmap(io, Vector{UInt32}, (1024,), 0, grow=false)
ERROR: IOError: filesize: Illegal seek for <file /dev/gpiomem>
Stacktrace:
 [1] filesize(s::IOStream)
   @ Base ./iostream.jl:226
 [2] mmap(io::IOStream, ::Type{Vector{UInt32}}, dims::Tuple{Int64}, offset::Int64; grow::Bool, shared::Bool)
   @ Mmap ~/.julia/juliaup/julia-1.10.1+0.aarch64.linux.gnu/share/julia/stdlib/v1.10/Mmap/src/Mmap.jl:215
 [3] top-level scope
   @ REPL[4]:1
```

Version 1.10.2 has the same issue, I have not tested on 1.10.3 release candidate.

### PR DESCRIPTION
**Pull request by IanButterworth**
Fixes https://github.com/JuliaLang/julia/issues/54203
Requires #55641

Based on https://github.com/JuliaLang/julia/pull/55641#issuecomment-2334162489
cc. @JakeZw @ronisbr

### CODE REVIEW

**Comment by ronisbr** [comment]
Oh, that was fast :) Thanks @IanButterworth ! LGTM!

@JakeZw can you test in your setup, please, applying the changes in #55641 and here?

**Comment by IanButterworth** [comment]
This won't be ready for testing until https://github.com/JuliaLang/julia/pull/55641 is merged and this rebased. (later today probably)

**Comment by IanButterworth** [comment]
Hitting an error
```
  | 2024-09-23 21:44:31 EDT | SharedArrays                                     (1) \|         failed at 2024-09-23T18:44:32.107
  | 2024-09-23 21:44:32 EDT | Error During Test at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-a7c648dd99/share/julia/test/testdefs.jl:24
  | 2024-09-23 21:44:32 EDT | Got exception outside of a @test
  | 2024-09-23 21:44:32 EDT | LoadError: stat returned zero type for a valid path
  | 2024-09-23 21:44:32 EDT | Stacktrace:
  | 2024-09-23 21:44:32 EDT | [1] error(s::String)
  | 2024-09-23 21:44:32 EDT | @ Base ./error.jl:44
  | 2024-09-23 21:44:32 EDT | [2] stat(fd::RawFD)
  | 2024-09-23 21:44:32 EDT | @ Base.Filesystem ./stat.jl:180
  | 2024-09-23 21:44:32 EDT | [3] stat
  | 2024-09-23 21:44:32 EDT | @ ./iostream.jl:63 [inlined]
  | 2024-09-23 21:44:32 EDT | [4] isfile
  | 2024-09-23 21:44:32 EDT | @ ./stat.jl:503 [inlined]
  | 2024-09-23 21:44:32 EDT | [5] mmap(io::IOStream, ::Type{Array{Int64, 3}}, dims::Tuple{Int64, Int64, Int64}, offset::Int64; grow::Bool, shared::Bool)
  | 2024-09-23 21:44:32 EDT | @ Mmap /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-a7c648dd99/share/julia/stdlib/v1.12/Mmap/src/Mmap.jl:224
  | 2024-09-23 21:44:32 EDT | [6] _shm_mmap_array(T::Type, dims::Tuple{Int64, Int64, Int64}, shm_seg_name::String, mode::UInt32)
  | 2024-09-23 21:44:32 EDT | @ SharedArrays /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-a7c648dd99/share/julia/stdlib/v1.12/SharedArrays/src/SharedArrays.jl:692
  | 2024-09-23 21:44:32 EDT | [7] shm_mmap_array(T::Type, dims::Tuple{Int64, Int64, Int64}, shm_seg_name::String, mode::UInt32)
  | 2024-09-23 21:44:32 EDT | @ SharedArrays /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-a7c648dd99/share/julia/stdlib/v1.12/SharedArrays/src/SharedArrays.jl:651
  | 2024-09-23 21:44:32 EDT | [8] SharedArray{Int64, 3}(dims::Tuple{Int64, Int64, Int64}; init::Function, pids::Vector{Int64})
  | 2024-09-23 21:44:32 EDT | @ SharedArrays /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-a7c648dd99/share/julia/stdlib/v1.12/SharedArrays/src/SharedArrays.jl:118
  | 2024-09-23 21:44:32 EDT | [9] SharedArray
  | 2024-09-23 21:44:32 EDT | @ /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-a7c648dd99/share/julia/stdlib/v1.12/SharedArrays/src/SharedArrays.jl:104 [inlined]
  | 2024-09-23 21:44:32 EDT | [10] #shmem_rand#73
  | 2024-09-23 21:44:32 EDT | @ /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-a7c648dd99/share/julia/stdlib/v1.12/SharedArrays/src/SharedArrays.jl:548 [inlined]
  | 2024-09-23 21:44:32 EDT | [11] shmem_rand(TR::UnitRange{Int64}, dims::Tuple{Int64, Int64, Int64})
  | 2024-09-23 21:44:32 EDT | @ SharedArrays /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-a7c648dd99/share/julia/stdlib/v1.12/SharedArrays/src/SharedArrays.jl:546
```

File: stdlib/Mmap/src/Mmap.jl [hunk 1]
```diff
@@ -45,6 +45,7 @@ Anonymous() = Anonymous("",false,true)
 Base.isopen(::Anonymous) = true
 Base.isreadable(::Anonymous) = true
 Base.iswritable(a::Anonymous) = !a.readonly
+Base.isfile(::Anonymous) = true
```
Comment (IanButterworth) [inline stdlib/Mmap/src/Mmap.jl:48]: Seems counter to the definition of Anonymous but I believe it's what it needs to be to pass the logic we're adding.
Perhaps there needs to be a `isresizable` or something in the API?

File: stdlib/Mmap/src/Mmap.jl [hunk 2]
```diff
@@ -218,7 +221,7 @@ function mmap(io::IO,
     # platform-specific mmapping
     @static if Sys.isunix()
         prot, flags, iswrite = settings(file_desc, shared)
-        if requestedSizeLarger
+        if isfile(io) && requestedSizeLarger  # add a condition to this line to ensure it only checks files
```
Comment (vtjnash) [inline stdlib/Mmap/src/Mmap.jl:224]:
```suggestion
        if requestedSizeLarger && isfile(io) # add a condition to this line to ensure it only checks files
```

File: stdlib/Mmap/src/Mmap.jl (hunk 1 above)
Comment (vtjnash) [reply stdlib/Mmap/src/Mmap.jl:48]:
```suggestion
```

**Review by vtjnash — COMMENTED** [review]
I think you just need to swap the order of the `requestedSizeLarger` check later, since it it already guards against being Mmap.Anonymous

**Comment by JakeZw** [comment]
@ronisbr asked me to test on my PI.  I will gladly do this, however my skills figuring out how to test are sorely lacking.  Please walk me through the details.

**Comment by IanButterworth** [comment]
Once CI is green here:
1. Install juliaup
2. `juliaup add pr55849`
3. `julia +pr55849`
4. Run test

**Comment by JakeZw** [comment]
For CI to be green do the 3 green squares and 1 red square at the top right of this page have to be all green?  (+5-3)

**Comment by JakeZw** [comment]
I need to know how I can tell that the building process is complete.

**Comment by JakeZw** [comment]
Maybe if I read just above this comment box it say Building Pending - 
STarted, it will tell me when it is complete, Sorry I just need to read!

**Comment by JakeZw** [comment]
The statements that were erroring. namely

```
        gpiomem_io  = open("/dev/gpiomem", "w+")
        gpiomem_map = Mmap.mmap(
            gpiomem_io,
            Vector{UInt32},
            (1024,),
            0,
            grow = false
        )
```
Now work!  

My application that depends upon BareMetalPI also works.  Thanks to you all

**Pull request merged**